cpp/include/raft/core/memory_type.hpp

```diff
@@ -49,4 +79,29 @@ auto constexpr memory_type_from_access()
 }
 
 }  // end namespace detail
+
+template <typename T>
+auto memory_type_from_pointer(T* ptr)
+{
+  auto result = memory_type::host;
+#ifndef RAFT_DISABLE_CUDA
+  auto* void_ptr = static_cast<void*>(nullptr);
+  if constexpr (std::is_const_v<T>) {
+    void_ptr = const_cast<void*>(static_cast<void const*>(ptr));
+  } else {
+    void_ptr = static_cast<void*>(ptr);
+  }
+  auto attrs = cudaPointerAttributes{};
+  RAFT_CUDA_TRY(cudaPointerGetAttributes(&attrs, void_ptr));
```

Comment:
Does this throw a compiler error if the pointer is not cast to `void`?

Reply:
Yes I think the cudaPointerGetAttributes requires a void pointer.

Reply:
A pointer should be able to be implicitly cast to `void*`. My question is if a compiler error occurs without the explicit cast.

Reply:
Nope, it doesn't throw a compiler error, but I generally prefer to avoid implicit casts unless there's a good reason for them. Not religious about that, though.